src/pages/experiments/[experimentId]/data-processing/configure-embedding/components/EmbeddingPreview.jsx

```diff
@@ -1,165 +1,137 @@
-  const embedding = useSelector((state) => state.embeddings);
-  const embeddingMethod = embeddingConfig?.configureEmbedding?.embeddingSettings?.method;
+  const [selectedPlot, setSelectedPlot] = useState('sample');
+  const [plot, setPlot] = useState(false);
+  const cellSets = useSelector((state) => state.cellSets);
 
-  const error = false;
+  const dispatch = useDispatch();
 
   const plots = {
     sample: {
-      title: 'Samples',
-      initialConfig: initialPlotConfigStates.embeddingCategorical,
-      specGenerator: generateEmbeddingCategoricalSpec,
+      title: 'Colored by Samples',
       imgSrc: plot1Pic,
+      plotUuid: 'embeddingPreviewBySample',
+      plotType: 'embeddingPreviewBySample',
     },
     cellCluster: {
-      title: 'Default clusters',
-      initialConfig: initialPlotConfigStates.embeddingCategorical,
-      specGenerator: generateEmbeddingCategoricalSpec,
+      title: 'Colored by CellSets',
       imgSrc: plot1Pic,
+      plotUuid: 'embeddingPreviewByCellSets',
+      plotType: 'embeddingPreviewByCellSets',
     },
     mitochondrialFraction: {
       title: 'Mitochondrial fraction reads',
-      initialConfig: initialPlotConfigStates.embeddingContinuous,
-      specGenerator: generateEmbeddingContinuousSpec,
       imgSrc: plot2Pic,
+      plotUuid: 'embeddingPreviewMitochondrialReads',
+      plotType: 'embeddingPreviewMitochondrialReads',
     },
     doubletScore: {
       title: 'Cell doublet score',
-      initialConfig: initialPlotConfigStates.embeddingContinuous,
-      specGenerator: generateEmbeddingContinuousSpec,
       imgSrc: plot2Pic,
+      plotUuid: 'embeddingPreviewDoubletScore',
+      plotType: 'embeddingPreviewDoubletScore',
     },
   };
 
+  const config = useSelector((state) => state.componentConfig[plots[selectedPlot].plotUuid]?.config);
+
+  useEffect(() => {
+    dispatch(loadCellSets(experimentId));
+  }, [experimentId]);
+
   useEffect(() => {
     // Do not update anything if the cell sets are stil loading or if
     // the config does not exist yet.
-
     if (!config) {
       return;
     }
 
-    const spec = plots[selectedSpec].specGenerator(config);
-    generateData(spec);
-    setPlotSpec(spec);
-  }, [config]);
+    if (!cellSets.loading && !cellSets.error) {
+      switch (plots[selectedPlot].plotType) {
+        case 'embeddingPreviewMitochondrialReads':
+        case 'embeddingPreviewDoubletScore':
+          setPlot(<ContinuousEmbeddingPlot experimentId={experimentId} config={config} plotUuid={plots[selectedPlot].plotUuid} />);
+          break;
+
+        // Set this as default plot because default plot is 'sample' which is categorical
+        case 'embeddingPreviewBySample':
+        case 'embeddingPreviewByCellSets':
+        default:
+          setPlot(<CategoricalEmbeddingPlot experimentId={experimentId} config={config} plotUuid={plots[selectedPlot].plotUuid} />);
+          break;
+      }
```

Comment:
you can avoid this switch statement if you add the component for each relevant view in the `plots` object, similarly to how we do this in data exploration: https://github.com/biomage-ltd/ui/blob/master/src/pages/experiments/%5BexperimentId%5D/data-exploration/index.jsx#L54

something like:

```
const plots = {
    sample: {
      title: 'Samples',
      initialConfig: initialPlotConfigStates.embeddingCategorical,
      specGenerator: generateEmbeddingCategoricalSpec,
      title: 'Colored by Samples',
      imgSrc: plot1Pic,
      plotUuid: 'embeddingPreviewBySample',
      plotType: 'embeddingPreviewBySample',
      component: <CategoricalEmbeddingPlot experimentId={experimentId} config={config} plotUuid={plots[selectedPlot].plotUuid} />
    },
```

and then in this code here all you need to do is `setPlot(plots[selectedPlot].component)`

Reply:
refactored in new commit. What's interesting is that I have to implement `component` as a function so that it is lazily evaluated. Otherwise, there would be initialization dependency between `plots` and config, as they depend on each other for initialization.

Reply:
Correct, that is how it is with React. Components are rendered when the component statement is called, and they will not update unless you re-instantiate them. The common way to do this is to wrap it in an anonymous function, such as `() => <Component />`.